File: backend/verify_fix.py
import urllib.request
import urllib.parse
import json
import ssl

# Ignore SSL (localhost)
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

BASE_URL = "http://localhost:8000/v1"

def login(username, password):
    url = f"{BASE_URL}/auth/login"
    data = json.dumps({"account": username, "password": password}).encode('utf-8')
    req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'}, method='POST')
    try:
        with urllib.request.urlopen(req, context=ctx) as response:
            res_json = json.loads(response.read().decode('utf-8'))
            return res_json['data']['token']
    except Exception as e:
        print(f"Login failed: {e}")
        try:
             if hasattr(e, 'read'): print(e.read().decode())
        except: pass
        return None

def upload_file(token):
    # Create dummy file content
    boundary = '----WebKitFormBoundary7MA4YWxkTrZu0gW'
    data = []
    data.append(f'--{boundary}')
    data.append('Content-Disposition: form-data; name="file"; filename="test.png"')
    data.append('Content-Type: image/png')
    data.append('')
    
    # Needs to be bytes
    body_parts = []
    for item in data:
        body_parts.append(item.encode('utf-8'))
    
    # Inject binary content for "fake image" if strict check, but text is fine?
    # backend checks startsWith("image/") content-type header of the part
    # file.content_type in fastapi comes from the header
    
    body_parts[-2] = b'fakeimagecontent' # replace empty string with content
    
    # Construct body
    final_body = b''
    for part in body_parts:
        final_body += part + b'\r\n'
    final_body += f'--{boundary}--\r\n'.encode('utf-8')
    
    
    url = f"{BASE_URL}/upload"
    req = urllib.request.Request(url, data=final_body, headers={
        'Authorization': f'Bearer {token}',
        'Content-Type': f'multipart/form-data; boundary={boundary}'
    }, method='POST')
    
    try:
        with urllib.request.urlopen(req, context=ctx) as response:
            res_json = json.loads(response.read().decode('utf-8'))
            print("Upload Success:", res_json)
            return True
    except urllib.error.HTTPError as e:
        print(f"Upload failed: {e.code} {e.read().decode('utf-8')}")
        return False
    except Exception as e:
        print(f"Upload error: {e}")
        return False

# Test
print("Attempting login as testuser...")
token = login("testuser", "test123")
if token:
    print("Logged in as testuser.")
    print("Attempting upload...")
    if upload_file(token):
        print("VERIFICATION PASSED: User can upload files.")
    else:
        print("VERIFICATION FAILED: User cannot upload files.")
else:
    print("Could not login. Trying admin to ensure server is up.")
    token = login("admin", "admin123")
    if token:
         print("Admin login worked, but testuser failed. Check init_data.")
    else:
         print("Server might be down or unreachable.")

File: backend/verify_fix.js
const http = require('http');

const register = () => {
    return new Promise((resolve, reject) => {
        const username = `test_${Date.now()}`;
        const email = `${username}@example.com`;
        const password = "password123";

        const req = http.request({
            hostname: 'localhost',
            port: 8000,
            path: '/v1/auth/register',
            method: 'POST',
            headers: { 'Content-Type': 'application/json' }
        }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    if (json.code === 200) {
                        console.log("Registration successful for", username);
                        resolve({ username, password });
                    }
                    else if (json.message && json.message.includes("exists")) {
                        // If exists, try login (fallback)
                        console.log("User exists, trying login...");
                        resolve({ username: "testuser", password: "test123" }); // Fallback to testuser if we can't register? 
                        // No, better to resolve with the created creds
                        reject(json.message);
                    }
                    else reject(json.message || "Registration failed");
                } catch (e) { reject(e); }
            });
        });
        req.on('error', reject);
        req.write(JSON.stringify({ username, email, password }));
        req.end();
    });
};

const login = (username, password) => {
    return new Promise((resolve, reject) => {
        const req = http.request({
            hostname: 'localhost',
            port: 8000,
            path: '/v1/auth/login',
            method: 'POST',
            headers: { 'Content-Type': 'application/json' }
        }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    if (json.code === 200) resolve(json.data.token);
                    else reject(json.message || "Login failed");
                } catch (e) { reject(e); }
            });
        });
        req.on('error', reject);
        req.write(JSON.stringify({ account: username, password: password }));
        req.end();
    });
};

const upload = (token) => {
    return new Promise((resolve, reject) => {
        const boundary = '----WebKitFormBoundary7MA4YWxkTrZu0gW';
        const content = 'fakeimagecontent';
        const body = `--${boundary}\r\n` +
            `Content-Disposition: form-data; name="file"; filename="test.png"\r\n` +
            `Content-Type: image/png\r\n\r\n` +
            `${content}\r\n` +
            `--${boundary}--\r\n`;

        const req = http.request({
            hostname: 'localhost',
            port: 8000,
            path: '/v1/upload',
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${token}`,
                'Content-Type': `multipart/form-data; boundary=${boundary}`,
                'Content-Length': Buffer.byteLength(body)
            }
        }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                console.log("Upload response:", data);
                if (res.statusCode === 200) {
                    try {
                        const json = JSON.parse(data);
                        if (json.code === 200) resolve();
                        else reject(json.message);
                    } catch (e) { reject("Invalid JSON"); }
                }
                else reject(`Status: ${res.statusCode}, Body: ${data}`);
            });
        });
        req.on('error', reject);
        req.write(body);
        req.end();
    });
};

(async () => {
    try {
        console.log("Registering...");
        const creds = await register();
        console.log("Logging in...");
        const token = await login(creds.username, creds.password);
        console.log("Logged in. Uploading...");
        await upload(token);
        console.log("VERIFICATION PASSED");
    } catch (e) {
        console.error("VERIFICATION FAILED:", e);
        process.exit(1);
    }
})();
